sys/proto.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * sys/proto.c
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Ziemowit Leszczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <netdb.h>
+#include <string.h>
+#include <stdlib.h>
+
+
+#define PROTOCOLS_NUM 55
+#define MAX_ALIASES   2
+
+
+struct protoent_internal {
+	const char *name;
+	const char *aliases[MAX_ALIASES];
+	int proto;
+};
+
+
+struct protoent_internal protocols[PROTOCOLS_NUM] = {
+	{ "ip", { "IP", NULL }, 0 },
+	{ "hopopt", { "HOPOPT", NULL }, 0 },
+	{ "icmp", { "ICMP", NULL }, 1 },
+	{ "igmp", { "IGMP", NULL }, 2 },
+	{ "ggp", { "GGP", NULL }, 3 },
+	{ "ipencap", { "IP-ENCAP", NULL }, 4 },
+	{ "st", { "ST", NULL }, 5 },
+	{ "tcp", { "TCP", NULL }, 6 },
+	{ "egp", { "EGP", NULL }, 8 },
+	{ "igp", { "IGP", NULL }, 9 },
+	{ "pup", { "PUP", NULL }, 12 },
+	{ "udp", { "UDP", NULL }, 17 },
+	{ "hmp", { "HMP", NULL }, 20 },
+	{ "xns-idp", { "XNS-IDP", NULL }, 22 },
+	{ "rdp", { "RDP", NULL }, 27 },
+	{ "iso-tp4", { "ISO-TP4", NULL }, 29 },
+	{ "dccp", { "DCCP", NULL }, 33 },
+	{ "xtp", { "XTP", NULL }, 36 },
+	{ "ddp", { "DDP", NULL }, 37 },
+	{ "idpr-cmtp", { "IDPR-CMTP", NULL }, 38 },
+	{ "ipv6", { "IPv6", NULL }, 41 },
+	{ "ipv6-route", { "IPv6-Route", NULL }, 43 },
+	{ "ipv6-frag", { "IPv6-Frag", NULL }, 44 },
+	{ "idrp", { "IDRP", NULL }, 45 },
+	{ "rsvp", { "RSVP", NULL }, 46 },
+	{ "gre", { "GRE", NULL }, 47 },
+	{ "esp", { "IPSEC-ESP", NULL }, 50 },
+	{ "ah", { "IPSEC-AH", NULL }, 51 },
+	{ "skip", { "SKIP", NULL }, 57 },
+	{ "ipv6-icmp", { "IPv6-ICMP", NULL }, 58 },
+	{ "ipv6-nonxt", { "IPv6-NoNxt", NULL }, 59 },
+	{ "ipv6-opts", { "IPv6-Opts", NULL }, 60 },
+	{ "rspf", { "RSPF", "CPHB" }, 73 },
+	{ "vmtp", { "VMTP", NULL }, 81 },
+	{ "eigrp", { "EIGRP", NULL }, 88 },
+	{ "ospf", { "OSPFIGP", NULL }, 89 },
+	{ "ax.25", { "AX.25", NULL }, 93 },
+	{ "ipip", { "IPIP", NULL }, 94 },
+	{ "etherip", { "ETHERIP", NULL }, 97 },
+	{ "encap", { "ENCAP", NULL }, 98 },
+	{ "pim", { "PIM", NULL }, 103 },
+	{ "ipcomp", { "IPCOMP", NULL }, 108 },
+	{ "vrrp", { "VRRP", NULL }, 112 },
+	{ "l2tp", { "L2TP", NULL }, 115 },
+	{ "isis", { "ISIS", NULL }, 124 },
+	{ "sctp", { "SCTP", NULL }, 132 },
+	{ "fc", { "FC", NULL }, 133 },
+	{ "mobility-header", { "Mobility-Header", NULL }, 135 },
+	{ "udplite", { "UDPLite", NULL }, 136 },
+	{ "mpls-in-ip", { "MPLS-in-IP", NULL }, 137 },
+	{ "manet", { NULL, NULL }, 138 },
+	{ "hip", { "HIP", NULL }, 139 },
+	{ "shim6", { "Shim6", NULL }, 140 },
+	{ "wesp", { "WESP", NULL }, 141 },
+	{ "rohc", { "ROHC", NULL }, 142 }
+};
+
+
+static char *aliases_common[MAX_ALIASES + 1];
+static struct protoent proto_common = { NULL, aliases_common, -1 };
+
+
+static void protoent_free(struct protoent *proto)
+{
+	unsigned int i = 0;
+
+	if (proto->p_name)
+		free(proto->p_name);
+
+	if (proto->p_aliases) {
+		while (i < MAX_ALIASES && proto->p_aliases[i]) {
```

Comment:
This while works exactly once - for `i == 0`. Is this intended? What about `p_aliases[1]` and `p_aliases[2]` (`p_aliases` have 3 indexes)?

Reply:
Also, `for` loop fits here better

Reply:
It loops 0 to MAX_ALIASES times. `p_aliases` is array of string pointers terminated by NULL. This function is used only on `proto_common` variable.